src/Microsoft.Diagnostics.Monitoring.WebApi/LoggingExtensions.cs

```diff
@@ -13,37 +13,37 @@ internal static class LoggingExtensions
             LoggerMessage.Define(
                 eventId: new EventId(1, "RequestFailed"),
                 logLevel: LogLevel.Error,
-                formatString: "Request failed.");
+                formatString: Strings.LogFormatString_RequestFailed);
 
         private static readonly Action<ILogger, Exception> _requestCanceled =
             LoggerMessage.Define(
                 eventId: new EventId(2, "RequestCanceled"),
                 logLevel: LogLevel.Information,
-                formatString: "Request canceled.");
+                formatString: Strings.LogFormatString_RequestCanceled);
 
         private static readonly Action<ILogger, Exception> _resolvedTargetProcess =
             LoggerMessage.Define(
                 eventId: new EventId(3, "ResolvedTargetProcess"),
                 logLevel: LogLevel.Debug,
-                formatString: "Resolved target process.");
+                formatString: Strings.LogFormatString_ResolvedTargetProcess);
 
         private static readonly Action<ILogger, string, Exception> _egressedArtifact =
             LoggerMessage.Define<string>(
                 eventId: new EventId(4, "EgressedArtifact"),
                 logLevel: LogLevel.Information,
-                formatString: "Egressed artifact to {location}");
+                formatString: Strings.LogFormatString_EgressedArtifact);
 
         private static readonly Action<ILogger, Exception> _writtenToHttpStream =
             LoggerMessage.Define(
                 eventId: new EventId(5, "WrittenToHttpStream"),
                 logLevel: LogLevel.Information,
-                formatString: "Written to HTTP stream.");
+                formatString: Strings.LogFormatString_WrittenToHttpStream);
 
         private static readonly Action<ILogger, int, int, Exception> _throttledEndpoint =
             LoggerMessage.Define<int, int>(
                  eventId: new EventId(6, "ThrottledEndpoint"),
                 logLevel: LogLevel.Warning,
-                formatString: "Request limit for endpoint reached. Limit: {limit}, oustanding requests: {requests}");
```

Comment:
I know you did a whole bunch of work here to make the log statements localizable, but there will be an issue if we go to actually localize these. The parameters in the log statement string are positional (e.g. the `limit` parameter is filled from the first argument, the `requests` parameter is filled from the second argument). When this statement is localized, the parameters may be switched around depending on the language. This will cause arguments to not be correctly matched with parameters. So I'm wondering if we should (a) not localize log statements, and/or (b) consult with someone in corefx about localization + logging.

Reply:
FYI @maryamariyan. Do we have established guidance for this question?

Reply:
Similar to APIs like [`TextWriter.WriteLine(string format, object? arg0, object? arg1)`](https://docs.microsoft.com/en-us/dotnet/api/system.io.textwriter.writeline?view=net-5.0#System_IO_TextWriter_WriteLine_System_String_System_Object_System_Object_), or [`string.Format(string format, object? arg0, object? arg1, object? arg2)`](https://docs.microsoft.com/en-us/dotnet/api/system.string.format?view=net-5.0) the action returning `LoggerMessage.Define<.., T1, .., TN, ..>` APIs format strings according to order in which they receive name hole values within the formatted string. 

Therefore, it is up to the consumer of these APIs to carefully decide how to use these APIs based on localization (perhaps to flip parameter orders given a specific localization).

/cc @tarekgh

Reply:
> Similar to APIs like [`TextWriter.WriteLine(string format, object? arg0, object? arg1)`](https://docs.microsoft.com/en-us/dotnet/api/system.io.textwriter.writeline?view=net-5.0#System_IO_TextWriter_WriteLine_System_String_System_Object_System_Object_), or [`string.Format(string format, object? arg0, object? arg1, object? arg2)`](https://docs.microsoft.com/en-us/dotnet/api/system.string.format?view=net-5.0) the action returning `LoggerMessage.Define<.., T1, .., TN, ..>` APIs format strings according to order in which they receive name hole values within the formatted string.

I don't think that is necessarily correct. Those APIs fill the format according to the parameter index defined within the format string. Because they use parameter index, it doesn't matter how the format string arranges them. https://docs.microsoft.com/en-us/dotnet/api/system.io.textwriter.writeline?view=net-5.0#System_IO_TextWriter_WriteLine_System_String_System_Object_System_Object_System_Object_ clearly states that it uses indexes, not names, for the parameters.

On the other hand, it is encouraged that the format string for ILogger uses named parameters (from which their positions are inferred). Thus, if a format string is localized and the positions of those named parameters change within the format string, then they will match to the wrong positions.

I'm not certain if ILogger supports using parameter indexes (my guess that it doesn't in the same sense as the above APIs, probably treats them as names), but then you lose fidelity of the parameter values when consuming the logs (e.g. via console logs, saying parameter "0" is "John" is less information than saying parameter "firstName" is "John").

Reply:
> Therefore, it is up to the consumer of these APIs to carefully decide how to use these APIs based on localization (perhaps to flip parameter orders given a specific localization).

I think it is unlikely that anyone will be able to take an arbitrarily localized string and to know at execution time what the new parameter order is and then dynamically adjust the parameter order. Localization (at least via resx) does not provide rich enough information of the parameter reordering.

I could hypothetically say "if Spanish, use this order; if Japanese, use this order" etc, but that is not scalable to every individual string and can easily change if some decides to relocalize the string in a slightly different order.

Given that, I think the conclusion is that localization is highly discouraged due to reordering of named parameters, but I would love to be wrong.

Reply:
It is impossible for the logging APIs to format according to the resource language. The reorder of the string parts is not easy to handle and need to know how the rendering of this string can be later. I can give tons of examples of that but the easiest one is when using BiDi languages which the order of the string parts can change if your render the whole string from right to left or render it from right to left. 

> Given that, I think the conclusion is that localization is highly discouraged due to reordering of named parameters, but I would love to be wrong.

I wouldn't say that though. The callers of the logging can do some testing with different languages and can decide if it is good enough or not. The guidelines should be users of the logging APIs need to do some localization testing.

Reply:
> It is impossible for the logging APIs to format according to the resource language.

As logging is today, yes.

However, it could support parameter indexes. I haven't thought it through, but my first pass would be to allow optional inclusion of the parameter index in the slots e.g. "The object has the color {0:color} and has the shape {1:shape}". Whether these two slots are rearranged in any particular language is then not a worry of either the developer nor the localizer. This allows explicit assignment of which parameter index to get the value from as well the name of the parameter when it is logged. I don't know what other capabilities are allowed for each slot (is format specifiers a thing for logger format strings?), but this would be my first naive attempt.

> The callers of the logging can do some testing with different languages and can decide if it is good enough or not. The guidelines should be users of the logging APIs need to do some localization testing.

Given the current functionality, assessing all of the localized strings for a non-trivial project would be difficult. I could write a test that extracts all of the localized content and checks whether the order of the slots are correct. But when localization team inserts a new update in the future that breaks that arrangement, my recourses are (a) to tell them "no, you can't localize it that way" but I don't think it's my place to dictate how a native language speaker expresses that content, or (b) special case each and every violation to rearrange the parameters into what the language expects on the fly at runtime and add that new arrangement to the test. I feel like both of those solutions are untenable, but that's for each team to assess or maybe there is another low manual maintenance solution that I haven't thought.

Given all of that, I would make the argument for the projects that I work on to never localize ILogger statements.

Reply:
> However, it could support parameter indexes.

This still wouldn't be enough to ensure the final resulted text would be in the right form for different languages. There is extreme complexity which usually handled by the rendering engines.  Imagine for a second the localized strings and the parameters are ended up in different languages and need to fit together. Look at [Unicode BiDi Algorithm](https://unicode.org/reports/tr9/) as example to get the sense of the difficulty to layout the resulted string in the right form.

Reply:
> > However, it could support parameter indexes.
> 
> This still wouldn't be enough to ensure the final resulted text would be in the right form for different languages.

How so? If using indexes, it should not be any different algorithmically than how `string.Format` does slot replacement.

I don't know what "rendering engines" have to do with this. All this is about is matching the correct arguments with the indexes as described by the format string. It does not have to have any knowledge of the language of the format string.